File: solutions/typescript/two-bucket/2/two-bucket.ts
function gcd(a: number, b: number): number {
  if (b === 0) {
    return a
  }
  return gcd(b, a % b)
}

export class TwoBucket {
  private _goalBucket = 'one'
  private _otherBucket = 0
  private _goal: number
  private _steps = 0
  private _starterBuck: boolean
  private _buckOne: number
  private _buckTwo: number

  constructor(buckOne: number, buckTwo: number, goal: number, starterBuck: string) {
    this._goal = goal
    this._starterBuck = starterBuck === 'one'
    this._buckOne = this._starterBuck ? buckOne : buckTwo
    this._buckTwo = this._starterBuck ? buckTwo : buckOne
  }

  moves() : number {
    if (this._goal % gcd(this._buckOne, this._buckTwo) !== 0) {
      throw new Error("Not possible to reach the goal ")
    }
    if (this._goal > this._buckOne && this._goal > this._buckTwo) {
      throw new Error("Not possible to reach the goal ")
    }

    if (this._goal === this._buckTwo) {
      this._goalBucket = 'two'
      this._otherBucket = this._buckOne
      return 2
    }

    let a = this._buckOne
    let b = 0

    this._steps += 1

    while (a !== this._goal && b !== this._goal) {
      const amt = Math.min(a, this._buckTwo - b)
      a -= amt
      b += amt
      this._steps += 1

      if (a === this._goal || b === this._goal) {
        break
      }

      if (a === 0) {
        a = this._buckOne
        this._steps += 1
      }

      if (b === this._buckTwo) {
        b = 0
        this._steps += 1
      }
    }

    this._goalBucket = a === this._goal 
    ? this._starterBuck ? 'one' : 'two'
    : this._starterBuck ? 'two' : 'one'

    this._otherBucket = a === this._goal ? b : a

    return this._steps
  }

  get goalBucket() {
    return this._goalBucket
  }

  get otherBucket() {
    return this._otherBucket
  }

}

File: solutions/go/two-bucket/2/two_bucket.go
package twobucket

import "errors"

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type bucket struct {
	name  string
	size  int
	level int
}

func bucketFromValues(name string, size int) *bucket {
	b := bucket{name: name, size: size, level: 0}
	return &b
}

func (buc bucket) room() int {
	return buc.size - buc.level
}

func (buc bucket) full() bool {
	return buc.level == buc.size
}

func (buc bucket) isEmpty() bool {
	return buc.level == 0
}

func (buc *bucket) fill() {
	buc.level = buc.size
}

func (buc *bucket) empty() {
	buc.level = 0
}

func (buc *bucket) pourInto(other *bucket) {
	amount := min(buc.level, other.room())
	buc.level -= amount
	other.level += amount
}

type resolved struct {
	status string
	target bucket
	other  bucket
}

func hitTarget(goal int, start bucket, other bucket) resolved {
	if start.level == goal {
		return resolved{"resolved", start, other}
	}
	if other.level == goal {
		return resolved{"resolved", other, start}
	}
	return resolved{"not", start, other}
}

func Solve(sizeBucketOne, sizeBucketTwo, goalAmount int, startBucket string) (string, int, int, error) {
	if goalAmount > max(sizeBucketOne, sizeBucketTwo) || goalAmount < 1 {
		return "None", sizeBucketOne, sizeBucketTwo, errors.New("Invalid goal amount")
	}
	if min(sizeBucketOne, sizeBucketTwo) < 1 {
		return "None", sizeBucketOne, sizeBucketTwo, errors.New("Invalid bucket size")
	}
	var start *bucket
	if startBucket == "one" {
		start = bucketFromValues("one", sizeBucketOne)
	} else {
		start = bucketFromValues("two", sizeBucketTwo)
	}
	var other *bucket
	if startBucket == "two" {
		other = bucketFromValues("one", sizeBucketOne)
	} else {
		other = bucketFromValues("two", sizeBucketTwo)
	}

	moves := 0

	if other.size == goalAmount {
		return other.name, 2, start.size, nil
	}

	for {
		resolution := hitTarget(goalAmount, *start, *other)
		if resolution.status == "not" {
			if start.isEmpty() {
				start.fill()
			} else if other.full() {
				other.empty()
			} else {
				start.pourInto(other)
			}

			moves += 1

			if start.full() && other.full() {
				return "None", sizeBucketOne, sizeBucketTwo, errors.New("No solution")
			}
		} else {
			return resolution.target.name, moves, resolution.other.level, nil
		}

	}

}
